### ISSUE TITLE
rospy topic connection fails if publisher restarts on same xml-rpc port

### ISSUE THREAD

**Issue by adamantivm** [ros/ros_comm#128]
When using a fixed xml-rpc port number for a ROS node which publishes a topic, if the node is restarted then topic listeners will not re-connect to the publisher to receive new messages, even when the publisher is following the appropriate topic registration procedure on start-up (and unregistration on shutdown).

This problem probably went unnoticed so far because until now each new ros process was randomly choosing their xml-rpc port.

The procedure I'm using for reproducing this is unfortunately not completely straight-forward. I'm using a lightly patched version of rospy which allows specifying an xml-rpc port from parameters: https://github.com/ros/ros_comm/pull/127. I have cloned and made this fork of ros_comm and configured a ROS workspace to use this version of rospy instead of the installed one from repos.

I built the following small test program which uses rospy to start a node on a fixed port number and publish a test message:

```
#!/usr/bin/python
import roslib; roslib.load_manifest('rospy')
import rospy
from std_msgs.msg import String
import sys

if len(sys.argv) > 1:
    port = int(sys.argv[1])
else:
    port = 33330

rospy.init_node('testport',port=port,log_level=1,tcpros_port=port+1)

publisher = rospy.Publisher('/testpub', String)
rospy.sleep(1)

r = rospy.Rate(2)
i = 0

while not rospy.is_shutdown():
    publisher.publish("Out goes msg %d" % i)
    r.sleep()
    i += 1

print "Shutting down, please wait"
publisher.unregister()
rospy.sleep(1)
```

In order to reproduce the issue, open a roscore and a rostopic echo, then start test.py, let it run for a little while, kill it and start it again.
In other words:

```
roscore
rostopic echo /testpub
./test.py
# wait 5 seconds and hit ^C
./test.py # for the second time
```

You will notice that the first time it runs, rostopic can echo the published messages but the second time (and subsequent times) it cannot, unless rostopic echo is restarted.

This may be related with issue #111 if unregistration is indeed failing

**Comment by adamantivm** [comment]
Looking at a network trace I can see that both times the publisher is making the appropriate registerPublisher xmlrpc call (and also unregisterPublisher when shutting down), but only the first time it gets a requestTopic xmlrpc call from the subscriber - the second time it is ignored

**Comment by adamantivm** [comment]
After some digging in the rospy code everything seems to indicate that Subscriber disconnection from unregistered Publishers is not implemented, is this possible?

When the subscriber receives the publisherUpdate that the master sends after the publisher unregistered, the code I see in https://github.com/ros/ros_comm/blob/groovy-devel/clients/rospy/src/rospy/impl/masterslave.py#L502 will cause it to ignore the update (which now comes with an empty list of publishers), thus the _TopicImpl will continue to think it is connected to this publisher uri.

Likewise the code in the RegManager seems to only be prepared to add new connections and not to remove existing ones that have been notified as non existent anymore - 
https://github.com/ros/ros_comm/blob/groovy-devel/clients/rospy/src/rospy/impl/registration.py#L287

Finally, when the receive_loop in the subscriber connected to a particular publisher is closed because the connection is finished, this condition is ignored and the corresponding _TopicImpl doesn't get notified that this connection is no longer valid - https://github.com/ros/ros_comm/blob/groovy-devel/clients/rospy/src/rospy/impl/tcpros_base.py#L714

I'm going to keep digging as I find time to look into this but I would greatly appreciate if the people that know more about this module and its history can give me some hints about the above decisions - it is most likely that there are good reasons why this is working the way it is and I am too new to see it. Also pointers to any code samples where topic disconnection is being successfully tested.

**Comment by adamantivm** [comment]
After thinking through my findings above I thought of a simple workaroundish solution which I submitted in pull request #129. I hope you find this reasonable.

**Comment by dirk-thomas** [comment]
Solved (and confirmed in https://github.com/ros/ros_comm/pull/129#commitcomment-2980385) with pull request #215.
